File: bindings/dart/lib/src/keywords.dart
// Keyword values. Dart has no union types, so every keyword the IR accepts is
// an enum here, and `apply` is the escape hatch for anything not yet modelled.

/// How wrapped flex lines are distributed on the cross axis.
enum AlignContent {
  flexStart('flex-start'),
  flexEnd('flex-end'),
  center('center'),
  stretch('stretch'),
  spaceBetween('space-between'),
  spaceAround('space-around'),
  spaceEvenly('space-evenly');

  const AlignContent(this.value);

  final String value;
}

/// Cross-axis alignment. Also the type of `alignSelf`.
enum AlignItems {
  flexStart('flex-start'),
  flexEnd('flex-end'),
  center('center'),
  stretch('stretch'),
  baseline('baseline');

  const AlignItems(this.value);

  final String value;
}

/// Main-axis distribution.
enum JustifyContent {
  flexStart('flex-start'),
  flexEnd('flex-end'),
  center('center'),
  spaceBetween('space-between'),
  spaceAround('space-around'),
  spaceEvenly('space-evenly');

  const JustifyContent(this.value);

  final String value;
}

/// The main axis of a container.
enum FlexDirection {
  row('row'),
  column('column'),
  rowReverse('row-reverse'),
  columnReverse('column-reverse');

  const FlexDirection(this.value);

  final String value;
}

/// Whether flex items wrap onto new lines.
enum FlexWrap {
  wrap('wrap'),
  noWrap('nowrap'),
  wrapReverse('wrap-reverse');

  const FlexWrap(this.value);

  final String value;
}

/// Whether width and height include padding and border.
enum BoxSizing {
  borderBox('border-box'),
  contentBox('content-box');

  const BoxSizing(this.value);

  final String value;
}

/// Writing direction.
enum TextDirection {
  ltr('ltr'),
  rtl('rtl');

  const TextDirection(this.value);

  final String value;
}

/// How a node participates in layout.
enum Display {
  none('none'),
  flex('flex'),
  contents('contents');

  const Display(this.value);

  final String value;
}

/// What happens to content past a node's box.
enum Overflow {
  visible('visible'),
  hidden('hidden'),
  scroll('scroll');

  const Overflow(this.value);

  final String value;
}

/// Positioning scheme.
enum Position {
  absolute('absolute'),
  relative('relative'),
  static('static');

  const Position(this.value);

  final String value;
}

/// Where a page break may or must fall.
enum PageBreakMode {
  before('before'),
  after('after'),
  avoid('avoid');

  const PageBreakMode(this.value);

  final String value;
}

/// The shape a corner radius produces.
enum Corner {
  cut('cut'),
  round('round');

  const Corner(this.value);

  final String value;
}

/// How a photo fills its box.
enum ScaleType {
  cover('cover'),
  fill('fill'),
  contain('contain');

  const ScaleType(this.value);

  final String value;
}

/// Roman or slanted.
enum FontStyle {
  normal('normal'),
  italic('italic'),
  oblique('oblique');

  const FontStyle(this.value);

  final String value;
}

/// The line-breaking algorithm.
enum LineBreakMode {
  greedy('greedy'),
  knuthPlass('knuth-plass');

  const LineBreakMode(this.value);

  final String value;
}

/// What a clipped paragraph ends with.
enum TextOverflow {
  clip('clip'),
  ellipsis('ellipsis');

  const TextOverflow(this.value);

  final String value;
}

/// Horizontal alignment inside a paragraph.
enum TextAlign {
  left('left'),
  right('right'),
  center('center'),
  justify('justify');

  const TextAlign(this.value);

  final String value;
}

/// Greedy wrapping, or ragged-edge balancing.
enum TextWrapMode {
  wrap('wrap'),
  balance('balance');

  const TextWrapMode(this.value);

  final String value;
}

/// How an open path ends.
enum StrokeCap {
  butt('butt'),
  round('round'),
  square('square');

  const StrokeCap(this.value);

  final String value;
}

/// How two path segments meet.
enum StrokeJoin {
  bevel('bevel'),
  miter('miter'),
  round('round');

  const StrokeJoin(this.value);

  final String value;
}

/// Which regions of a self-intersecting path are inside it.
enum FillRule {
  evenOdd('evenodd'),
  nonZero('nonzero');

  const FillRule(this.value);

  final String value;
}

/// The paragraph's base direction for bidi resolution.
enum BaseDirection {
  ltr('ltr'),
  rtl('rtl'),
  auto('auto');

  const BaseDirection(this.value);

  final String value;
}

/// Whether the final page is full height or shrinks to its content.
enum LastPageHeight {
  uniform('uniform'),
  content('content');

  const LastPageHeight(this.value);

  final String value;
}

/// The granularity of the boxes `metadata` returns.
enum Granularity {
  node('node'),
  line('line'),
  word('word');

  const Granularity(this.value);

  final String value;
}

File: bindings/swift/Sources/Sone/Keywords.swift
// Keyword values are enums with String raw values, so a call site reads as
// `.spaceBetween` with no type to name and no string to misspell.

/// How wrapped flex lines are distributed on the cross axis.
public enum AlignContent: String {
    case flexStart = "flex-start"
    case flexEnd = "flex-end"
    case center = "center"
    case stretch = "stretch"
    case spaceBetween = "space-between"
    case spaceAround = "space-around"
    case spaceEvenly = "space-evenly"
}

/// Cross-axis alignment. Also the type of `alignSelf`.
public enum AlignItems: String {
    case flexStart = "flex-start"
    case flexEnd = "flex-end"
    case center = "center"
    case stretch = "stretch"
    case baseline = "baseline"
}

/// Main-axis distribution.
public enum JustifyContent: String {
    case flexStart = "flex-start"
    case flexEnd = "flex-end"
    case center = "center"
    case spaceBetween = "space-between"
    case spaceAround = "space-around"
    case spaceEvenly = "space-evenly"
}

/// The main axis of a container.
public enum FlexDirection: String {
    case row = "row"
    case column = "column"
    case rowReverse = "row-reverse"
    case columnReverse = "column-reverse"
}

/// Whether flex items wrap onto new lines.
public enum FlexWrap: String {
    case wrap = "wrap"
    case nowrap = "nowrap"
    case wrapReverse = "wrap-reverse"
}

/// Whether width and height include padding and border.
public enum BoxSizing: String {
    case borderBox = "border-box"
    case contentBox = "content-box"
}

/// Writing direction.
public enum WritingDirection: String {
    case ltr = "ltr"
    case rtl = "rtl"
}

/// How a node participates in layout.
public enum Display: String {
    case none = "none"
    case flex = "flex"
    case contents = "contents"
}

/// What happens to content past a node's box.
public enum Overflow: String {
    case visible = "visible"
    case hidden = "hidden"
    case scroll = "scroll"
}

/// Positioning scheme.
public enum Position: String {
    case absolute = "absolute"
    case relative = "relative"
    case `static` = "static"
}

/// Where a page break may or must fall.
public enum PageBreakMode: String {
    case before = "before"
    case after = "after"
    case avoid = "avoid"
}

/// The shape a corner radius produces.
public enum Corner: String {
    case cut = "cut"
    case round = "round"
}

/// How a photo fills its box.
public enum ScaleType: String {
    case cover = "cover"
    case fill = "fill"
    case contain = "contain"
}

/// Roman or slanted.
public enum FontStyle: String {
    case normal = "normal"
    case italic = "italic"
    case oblique = "oblique"
}

/// The line-breaking algorithm.
public enum LineBreakMode: String {
    case greedy = "greedy"
    case knuthPlass = "knuth-plass"
}

/// What a clipped paragraph ends with.
public enum TextOverflow: String {
    case clip = "clip"
    case ellipsis = "ellipsis"
}

/// Horizontal alignment inside a paragraph.
public enum TextAlign: String {
    case left = "left"
    case right = "right"
    case center = "center"
    case justify = "justify"
}

/// Greedy wrapping, or ragged-edge balancing.
public enum TextWrapMode: String {
    case wrap = "wrap"
    case balance = "balance"
}

/// How an open path ends.
public enum StrokeCap: String {
    case butt = "butt"
    case round = "round"
    case square = "square"
}

/// How two path segments meet.
public enum StrokeJoin: String {
    case bevel = "bevel"
    case miter = "miter"
    case round = "round"
}

/// Which regions of a self-intersecting path are inside it.
public enum FillRule: String {
    case evenOdd = "evenodd"
    case nonZero = "nonzero"
}

/// The paragraph's base direction for bidi resolution.
public enum BaseDirection: String {
    case ltr = "ltr"
    case rtl = "rtl"
    case auto = "auto"
}

/// Whether the final page is full height or shrinks to its content.
public enum LastPageHeight: String {
    case uniform = "uniform"
    case content = "content"
}

/// The granularity of the boxes `metadata` returns.
public enum Granularity: String {
    case node = "node"
    case line = "line"
    case word = "word"
}

/// The output formats the engine can encode.
public enum OutputFormat: String {
    case png = "png"
    case jpeg = "jpeg"
    case webp = "webp"
    case raw = "raw"
    case pdf = "pdf"
    case svg = "svg"

    /// The `SoneFormat` discriminant the C ABI expects.
    var code: Int32 {
        switch self {
        case .png: return 0
        case .jpeg: return 1
        case .webp: return 2
        case .raw: return 3
        case .pdf: return 4
        case .svg: return 5
        }
    }
}
